app/src/main/java/org/sagebionetworks/research/mpower/profile/MpProfileManager.kt

```diff
@@ -0,0 +1,130 @@
+import androidx.lifecycle.MediatorLiveData
+import io.reactivex.Single
+import org.sagebionetworks.bridge.android.manager.AuthenticationManager
+import org.sagebionetworks.bridge.android.manager.models.ProfileDataManager
+import org.sagebionetworks.bridge.android.manager.models.ProfileDataSource
+import org.sagebionetworks.bridge.rest.model.StudyParticipant
+import org.sagebionetworks.research.sageresearch.dao.room.AppConfigRepository
+import org.sagebionetworks.research.sageresearch.dao.room.ReportEntity
+import org.sagebionetworks.research.sageresearch.dao.room.ReportRepository
+import org.sagebionetworks.research.sageresearch.profile.ProfileDataLoader
+
+/**
+ * The ProfileManager loads ProfileDataSources and a ProfileDataManager from AppConfig, useing the data
+ * from these it loads the necessary Reports, StudyParticipant and returns a ProfileDataLoader.
+ */
+class MpProfileManager(
+        val reportRepo: ReportRepository,
+        val appConfigRepo: AppConfigRepository,
+        val authManager: AuthenticationManager) {
+
+
+    fun loadProfileDataSources() : LiveData<Map<String, ProfileDataSource>> {
+        val profileDataSourceLiveData = LiveDataReactiveStreams.fromPublisher(
+                appConfigRepo.profileDataSources.toFlowable())
+        return profileDataSourceLiveData
+    }
+
+    private fun loadDefaultProfileDataManager() : Single<ProfileDataManager> {
+        return appConfigRepo.profileDataManagers.map {
+            it.get("ProfileManager")
+        }
+    }
+
+    private fun loadDefaultProfileManagerLiveData() : LiveData<ProfileDataManager> {
+        return LiveDataReactiveStreams.fromPublisher(loadDefaultProfileDataManager().toFlowable())
+    }
+
+    private fun loadReports(reportKeys: Set<String>): LiveData<Map<String, ReportEntity?>> {
+        val mediator = MediatorLiveData<Map<String, ReportEntity?>>().apply {
+            val numSources = reportKeys.size
+            val reportMap: MutableMap<String, ReportEntity?> = mutableMapOf()
+
+            fun update() {
+                if (numSources == reportMap.size) {
+                    this.value = reportMap
+                }
+            }
+
+            for (key in reportKeys) {
+                val reportLiveData = reportRepo.fetchMostRecentReport(key)
+                addSource(reportLiveData) {
+                    val report = it?.firstOrNull()
+                    reportMap.put(key, report)
+                    update()
+                }
+
+            }
+
+        }
+        return mediator
+    }
+
+    fun profileDataLoader(): LiveData<ProfileDataLoader> {
+        val mediator = MediatorLiveData<ProfileDataLoader>().apply {
+            var profileDataManager: ProfileDataManager? = null
+            var reportMap: Map<String, ReportEntity?>? = null
+
+            fun update() {
+                if (profileDataManager != null && reportMap != null) {
+                    val user = authManager.userSessionInfo
```

Comment:
@liujoshua @nategbrown9 here is the actual fix.  I just pulled the study participant info from the cached user session in the auth manager instead.  It really doesn't make any sense to go to the web for this stuff anyways, because if the user is at a point in the app where they need profile data, the user session info will be populated anyways.

Reply:
The profile data is editable, so I don't think this will work. The user session info is only updated when a user signs in or is re-authenticated. So if a user updates firstName or something else on StudyParticipant, they will get weird behavior. Initially they will see their change since the ProfileViewModel updates the cachedProfileDataLoader with the new values. Once the ProfileViewModel is destroyed, if they come back to the profile screen it would show outdated values from the user session until a re-auth occurs.

Reply:
Good thoughts, you are definitely correct.

I didn't really like my fix, to keep a light-weight cached version of the study participant up to date and load from that initially as well if it's available, but it seems to work well.

See video of me changing the user's name, leaving the app, and coming back, and it retains its value below.  I also tested all the other profile functionality, and it works well.

https://user-images.githubusercontent.com/5590748/125152583-3af12b00-e102-11eb-8212-9ec23d8de0d7.mov